### PR TITLE
[IPC] Service call supporting custom protobuf types

### PR DESCRIPTION
**Pull request by floriantschopp**
# Description
Extend the service calls to take arbitrary (serializable) request and reply types.

## Type of change
Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)

## Checklist before requesting a review
- [x] I have performed a self-review of my code.
- [x] If this is a new component I have added examples.

### CODE REVIEW

File: modules/ipc/include/hephaestus/ipc/zenoh/service.h [hunk 1]
```diff
@@ -4,30 +4,142 @@
...
-  std::string value;
-};
-
+template <typename RequestT, typename ReplyT>
 class Service {
 public:
-  using Callback = std::function<std::string(const QueryRequest&)>;
-  Service(SessionPtr session, std::string topic, Callback&& callback);
+  using Callback = std::function<ReplyT(const RequestT&)>;
+  explicit Service(SessionPtr session, TopicConfig topic_config, Callback&& callback);
```
[the remark this answers is not in the document]
Comment (floriantschopp) [reply modules/ipc/include/hephaestus/ipc/zenoh/service.h:23]: sorry, leftover of previous trials

File: modules/ipc/include/hephaestus/ipc/zenoh/service.h [hunk 2]
```diff
@@ -4,30 +4,146 @@
...
+    return static_cast<std::string>(query.get_value().as_string_view());
+  } else {
+    RequestT request;
+    auto payload = query.get_value().get_payload();
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+    std::span<const std::byte> buffer(reinterpret_cast<const std::byte*>(payload.start), payload.len);
+    serdes::deserialize<RequestT>(buffer, request);
+    return request;
+  }
+}
```
Comment (filippobrizzi) [inline modules/ipc/include/hephaestus/ipc/zenoh/service.h:60]: let's wrap this in an `internal` namespace

File: modules/ipc/include/hephaestus/ipc/zenoh/service.h [hunk 3]
```diff
@@ -4,30 +4,146 @@
...
+  : session_(std::move(session)), topic_config_(std::move(topic_config)), callback_(std::move(callback)) {
+  checkTemplatedTypes<RequestT, ReplyT>();
+
+  auto query = [this](const zenohc::Query& query) mutable {
+    DLOG(INFO) << fmt::format("Received query from '{}'", query.get_keyexpr().as_string_view());
+
+    auto reply = this->callback_(createRequest<RequestT>(query));
+    zenohc::QueryReplyOptions options;
+    if constexpr (std::is_same_v<ReplyT, std::string>) {
+      options.set_encoding(zenohc::Encoding{ Z_ENCODING_PREFIX_APP_CUSTOM });  // Update encoding.
```
Comment (filippobrizzi) [inline modules/ipc/include/hephaestus/ipc/zenoh/service.h:73]:
```suggestion
      options.set_encoding(zenohc::Encoding{ Z_ENCODING_PREFIX_TEXT_PLAIN });  // Update encoding.
```

File: modules/ipc/include/hephaestus/ipc/zenoh/service.h [hunk 4]
```diff
@@ -4,30 +4,146 @@
...
+      query.reply(this->topic_config_.name, reply, options);
+    } else {
+      options.set_encoding(zenohc::Encoding{ Z_ENCODING_PREFIX_TEXT_PLAIN });  // Update encoding.
```
Comment (filippobrizzi) [inline modules/ipc/include/hephaestus/ipc/zenoh/service.h:76]:
```suggestion
      options.set_encoding(zenohc::Encoding{ Z_ENCODING_PREFIX_APP_CUSTOM });  // Update encoding.
```

File: modules/ipc/include/hephaestus/ipc/zenoh/service.h [hunk 5]
```diff
@@ -4,30 +4,146 @@
...
+template <typename RequestT, typename ReplyT>
+auto callService(const SessionPtr& session, const TopicConfig& topic_config, const RequestT& request,
+                 const std::chrono::milliseconds& timeout) -> std::optional<ReplyT> {
+  checkTemplatedTypes<RequestT, ReplyT>();
+
+  DLOG(INFO) << fmt::format("Calling service on '{}'", topic_config.name);
+  std::mutex m;
+  std::condition_variable done_signal;
+  bool done = false;
+  ReplyT reply_message;
```
Comment (filippobrizzi) [inline modules/ipc/include/hephaestus/ipc/zenoh/service.h:94]: I think this needs to be a vector as we could receive multiple response if there are multiple services with the same prefix as zenoh allows for query prefixes.

File: modules/ipc/include/hephaestus/ipc/zenoh/service.h [hunk 6]
```diff
@@ -4,30 +4,146 @@
...
+    DLOG(INFO) << fmt::format("Received answer of '{}'", sample->get_keyexpr().as_string_view());
+    if constexpr (std::is_same_v<ReplyT, std::string>) {
+      std::unique_lock<std::mutex> lock(m);
+      reply_message = sample->get_payload().as_string_view();
+    } else {
+      auto payload = sample->get_payload();
+      // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+      std::span<const std::byte> buffer(reinterpret_cast<const std::byte*>(payload.start), payload.len);
+      std::unique_lock<std::mutex> lock(m);
+      serdes::deserialize(buffer, reply_message);
```
Comment (filippobrizzi) [inline modules/ipc/include/hephaestus/ipc/zenoh/service.h:113]: we would need to find a way to check that the type match. At least check that the encoding is not text. Ideally we would want to send the type as well.

File: modules/ipc/include/hephaestus/ipc/zenoh/service.h [hunk 7]
```diff
@@ -4,30 +4,146 @@
...
+  if constexpr (std::is_same_v<RequestT, std::string>) {
+    options.set_value(zenohc::Value(request, Z_ENCODING_PREFIX_TEXT_PLAIN));
+  } else {
+    options.set_value(zenohc::Value(serdes::serialize(request), Z_ENCODING_PREFIX_APP_CUSTOM));
+  }
+  const auto success =
+      session->zenoh_session.get(topic_config.name.c_str(), "", { on_reply, on_done }, options, error_code);
+
+  std::unique_lock lock(m);
+  if (!done_signal.wait_for(lock, timeout, [&done] { return done; })) {
```
Comment (filippobrizzi) [inline modules/ipc/include/hephaestus/ipc/zenoh/service.h:136]: I think you don't need to have `done`
```suggestion
  if (done_signal.wait_for(lock, timeout) == std::cv_status:: timeout) {
```

File: modules/ipc/include/hephaestus/ipc/zenoh/service.h [hunk 8]
```diff
@@ -4,30 +4,146 @@
...
+}
+
+template <typename RequestT, typename ReplyT>
+auto callService(const SessionPtr& session, const TopicConfig& topic_config, const RequestT& request,
+                 const std::chrono::milliseconds& timeout) -> std::optional<ReplyT> {
+  checkTemplatedTypes<RequestT, ReplyT>();
+
+  DLOG(INFO) << fmt::format("Calling service on '{}'", topic_config.name);
+  std::mutex m;
+  std::condition_variable done_signal;
```
Comment (filippobrizzi) [inline modules/ipc/include/hephaestus/ipc/zenoh/service.h:92]: I haven't tried myself yet, but this could be a good use case to use `absl::Mutex` which comprise of a mutex plus a condition variable internally so we do not need two separate variable and all locking overhead

File: modules/examples/examples/zenoh_service_server.cpp [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+    const std::string topic = "test";
+    heph::ipc::zenoh::Service<heph::examples::types::Pose, heph::examples::types::Pose> server(
+        session, topic_config, callback);
+
+    LOG(INFO) << fmt::format("Server started. Wating for queries on {} topic", topic);
+
+    while (true) {
+      std::this_thread::sleep_for(std::chrono::seconds(1));
+    }
```
[the remark this answers is not in the document]
Comment (floriantschopp) [reply modules/examples/examples/zenoh_service_server.cpp:43]: I just took this from the other examples. Should we refactor all of them?

File: modules/ipc/include/hephaestus/ipc/zenoh/service.h (hunk 8 above)
Comment (floriantschopp) [reply modules/ipc/include/hephaestus/ipc/zenoh/service.h:102]: I don't think so, `absl::Mutex::Wait` still needs a condition if I understand correctly, so not sure if it would make things easier.

File: modules/ipc/include/hephaestus/ipc/zenoh/service.h (hunk 6 above)
Comment (floriantschopp) [reply modules/ipc/include/hephaestus/ipc/zenoh/service.h]: Added a check for the encoding.

File: modules/examples/examples/zenoh_service_client.cpp [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    LOG(INFO) << fmt::format("Calling service on topic: {} with {}.", topic_config.name,
+                             heph::examples::types::toString(query));
+    const auto replies =
+        heph::ipc::zenoh::callService<heph::examples::types::Pose, heph::examples::types::Pose>(
+            session, topic_config, query, K_TIMEOUT);
+    if (!replies.empty()) {
+      std::string reply_str;
+      std::for_each(replies.begin(), replies.end(), [reply_str](const auto& reply) {
+        return fmt::format("{}\n-\t {}", reply_str, heph::examples::types::toString(reply.value));
+      });
```
Comment (filippobrizzi) [inline modules/examples/examples/zenoh_service_client.cpp:41]:
```suggestion
      std::for_each(replies.begin(), replies.end(), [&reply_str](const auto& reply) {
        reply_str += fmt::format("\t {}\n", reply.value);
      });
```

File: modules/examples/examples/zenoh_string_service_server.cpp [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+      return reply;
+    };
+
+    heph::ipc::zenoh::Service<std::string, std::string> server(session, topic_config, callback);
+
+    LOG(INFO) << fmt::format("String server started. Wating for queries on '{}' topic", topic_config.name);
+
+    while (true) {
+      std::this_thread::sleep_for(std::chrono::seconds(1));
+    }
```
Comment (filippobrizzi) [inline modules/examples/examples/zenoh_string_service_server.cpp:40]: use std::atomic_flag

File: modules/ipc/include/hephaestus/ipc/zenoh/service.h [hunk 12]
```diff
@@ -4,30 +4,187 @@
...
+      auto payload = sample->get_payload();
+      // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+      std::span<const std::byte> buffer(reinterpret_cast<const std::byte*>(payload.start), payload.len);
+      std::unique_lock<std::mutex> lock(m);
+      ReplyT reply_deserialized;
+      serdes::deserialize(buffer, reply_deserialized);
+      reply_messages.emplace_back(
+          ServiceResponse<ReplyT>{ .topic = topic_config.name, .value = std::move(reply_deserialized) });
+    }
+  };
```
Comment (filippobrizzi) [inline modules/ipc/include/hephaestus/ipc/zenoh/service.h:147]: let's extract this into a function

File: modules/ipc/include/hephaestus/ipc/zenoh/service.h [hunk 13]
```diff
@@ -4,30 +4,187 @@
...
+
+  auto on_done = [&m, &done_signal]() {
+    std::lock_guard lock(m);
+    done_signal.notify_all();
+  };
+
+  zenohc::ErrNo error_code{};
```
Comment (filippobrizzi) [inline modules/ipc/include/hephaestus/ipc/zenoh/service.h:154]: move this right before `session.get`

File: modules/ipc/include/hephaestus/ipc/zenoh/service.h [hunk 14]
```diff
@@ -4,30 +4,187 @@
...
+      reply_messages.emplace_back(ServiceResponse<ReplyT>{
+          .topic = topic_config.name,
+          .value = static_cast<std::string>(sample->get_payload().as_string_view()) });
+    } else {
+      CHECK(sample->get_encoding() == Z_ENCODING_PREFIX_EMPTY)
+          << "Encoding for binary types should be Z_ENCODING_PREFIX_EMPTY";
+      auto payload = sample->get_payload();
+      // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+      std::span<const std::byte> buffer(reinterpret_cast<const std::byte*>(payload.start), payload.len);
+      std::unique_lock<std::mutex> lock(m);
```
Comment (filippobrizzi) [inline modules/ipc/include/hephaestus/ipc/zenoh/service.h:141]: you can actually move the lock right before the emplace

File: modules/ipc/include/hephaestus/ipc/zenoh/service.h [hunk 15]
```diff
@@ -4,30 +4,187 @@
...
+      ReplyT reply_deserialized;
+      serdes::deserialize(buffer, reply_deserialized);
+      reply_messages.emplace_back(
+          ServiceResponse<ReplyT>{ .topic = topic_config.name, .value = std::move(reply_deserialized) });
```
Comment (filippobrizzi) [inline modules/ipc/include/hephaestus/ipc/zenoh/service.h:145]: To get full benefit from the emplace
```suggestion
      reply_messages.emplace_back( topic_config.name, std::move(reply_deserialized));
```

File: modules/examples/examples/zenoh_service_client.cpp (hunk 10 above)
Comment (floriantschopp) [reply modules/examples/examples/zenoh_service_client.cpp:41]: we need the `toString` here, otherwise makes sense.

Comment (filippobrizzi) [reply modules/examples/examples/zenoh_service_client.cpp:41]: Pose supports formatter so you don't need the `toString`

File: modules/ipc/include/hephaestus/ipc/zenoh/service.h [hunk 16]
```diff
@@ -4,30 +4,203 @@
...
+    -> std::vector<ServiceResponse<ReplyT>> {
+  internal::checkTemplatedTypes<RequestT, ReplyT>();
+
+  LOG(INFO) << fmt::format("Calling service on '{}'", topic_config.name);
+  std::mutex m;
+  std::condition_variable done_signal;
+  std::vector<ServiceResponse<ReplyT>> reply_messages;
+
+  auto on_reply = [&reply_messages, &m, &topic_config](zenohc::Reply&& reply) {
+    internal::onReply(std::move(reply), reply_messages, m, topic_config.name);
```
Comment (filippobrizzi) [inline modules/ipc/include/hephaestus/ipc/zenoh/service.h:159]:
```suggestion
    internal::onReply(std::forward<zneohc::Reply>(reply), reply_messages, m, topic_config.name);
```

File: modules/ipc/include/hephaestus/ipc/zenoh/service.h [hunk 17]
```diff
@@ -4,30 +4,203 @@
...
+template <typename RequestT, typename ReplyT>
+Service<RequestT, ReplyT>::Service(SessionPtr session, TopicConfig topic_config, Callback&& callback)
+  : session_(std::move(session)), topic_config_(std::move(topic_config)), callback_(std::move(callback)) {
+  internal::checkTemplatedTypes<RequestT, ReplyT>();
+
+  auto query = [this](const zenohc::Query& query) mutable {
+    LOG(INFO) << fmt::format("Received query from '{}'", query.get_keyexpr().as_string_view());
+
+    auto reply = this->callback_(internal::deserializeRequest<RequestT>(query));
+    DLOG(INFO) << fmt::format("Replying to '{}'", query.get_keyexpr().as_string_view());
```
Comment (filippobrizzi) [inline modules/ipc/include/hephaestus/ipc/zenoh/service.h:129]: you are not printing any new information wrt line 126 so I would remove this line

File: modules/examples/examples/zenoh_service_server.cpp [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+        .orientation = Eigen::Quaterniond{ 1., 0.1, 0.2, 0.3 },  // NOLINT
+        .position = Eigen::Vector3d{ 1, 2, 3 },
+      };
+      return sample_reply;
+    };
+
+    heph::ipc::zenoh::Service<heph::examples::types::Pose, heph::examples::types::Pose> server(
+        session, topic_config, callback);
+
+    LOG(INFO) << fmt::format("Server started. Wating for queries on '{}' topic", topic_config.name);
```
[the remark this answers is not in the document]
Comment (floriantschopp) [reply modules/examples/examples/zenoh_service_server.cpp:50]: I do agree that i would much prefer that :) do you think it makes sense to add it as part of this PR or a later cleanup?

File: modules/examples/examples/CMakeLists.txt [hunk 19]
```diff
@@ -9,6 +9,34 @@ define_module_example(
...
+        PUBLIC_INCLUDE_PATHS
+        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/../include>
+        PUBLIC_LINK_LIBS hephaestus::ipc)
+
+define_module_example(
+        NAME zenoh_string_service_client
+        SOURCES zenoh_string_service_client.cpp zenoh_program_options.h
+        PUBLIC_INCLUDE_PATHS
+        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/../include>
+        PUBLIC_LINK_LIBS hephaestus::ipc)
```
[the remark this answers is not in the document]
Comment (floriantschopp) [reply modules/examples/examples/CMakeLists.txt:31]: already exists as a more general usage app here
`modules/ipc/apps/zenoh_string_service.cpp`

File: modules/examples/examples/zenoh_string_service_client.cpp [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+#include <zenohc.hxx>
+
+#include "hephaestus/examples/types/pose.h"
+#include "hephaestus/examples/types_protobuf/pose.h"
+#include "hephaestus/ipc/zenoh/service.h"
+#include "hephaestus/ipc/zenoh/session.h"
+#include "hephaestus/serdes/serdes.h"
+#include "zenoh_program_options.h"
+
+auto main(int argc, const char* argv[]) -> int {
```
[the remark this answers is not in the document]
Comment (floriantschopp) [reply modules/examples/examples/zenoh_string_service_client.cpp:19]: we already have both, one as an example that just runs, and one as a cli app, wdyt?

File: modules/ipc/apps/zenoh_string_service.cpp [hunk 21]
```diff
@@ -4,7 +4,7 @@
 
 #include <fmt/core.h>
 
-#include "hephaestus/ipc/zenoh/query.h"
+#include "hephaestus/ipc/zenoh/service.h"
```
[the remark this answers is not in the document]
Comment (floriantschopp) [reply modules/ipc/apps/zenoh_string_service.cpp:7]: see above

File: modules/ipc/include/hephaestus/ipc/zenoh/service.h (hunk 16 above)
Comment (filippobrizzi) [reply modules/ipc/include/hephaestus/ipc/zenoh/service.h:156]: in this case you are right it shouldn't matter, but in general you don't know the type of `reply` as it could be an lvalue so for intermediate step you use forward and not worry

Comment (filippobrizzi) [reply modules/ipc/include/hephaestus/ipc/zenoh/service.h:156]: ah interesting, ok no letr's keep std::move

File: modules/examples/examples/zenoh_service_server.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (filippobrizzi) [reply modules/examples/examples/zenoh_service_server.cpp:50]: we can clean up later, let's decide it properly

File: modules/examples/examples/zenoh_sub.cpp [hunk 22]
```diff
@@ -19,30 +21,41 @@
...
 
     auto session = heph::ipc::zenoh::createSession(std::move(session_config));
 
     auto cb = [topic = topic_config.name](const heph::ipc::MessageMetadata& metadata,
                                           const std::shared_ptr<heph::examples::types::Pose>& pose) {
-      fmt::println(">> Time: {}. Topic {}. From: {}. Counter: {}. Received {}",
-                   std::chrono::system_clock::time_point{
-                       std::chrono::duration_cast<std::chrono::system_clock::duration>(metadata.timestamp) },
-                   metadata.topic, metadata.sender_id, metadata.sequence_id, *pose);
+      LOG(INFO) << fmt::format(
```
Comment (filippobrizzi) [inline modules/examples/examples/zenoh_sub.cpp:48]: nit. I think this is the output of the example so I would keep it as `fmt::println`

File: modules/examples/examples/zenoh_sub.cpp [hunk 23]
```diff
@@ -19,30 +21,41 @@
 #include "hephaestus/ipc/zenoh/subscriber.h"
 #include "zenoh_program_options.h"
 
+namespace {
+volatile std::atomic_bool keep_running(true);  // NOLINT
```
Comment (filippobrizzi) [inline modules/examples/examples/zenoh_sub.cpp:25]:
```suggestion
std::atomic_flag stop_flag = false;  // NOLINT
```

File: modules/examples/examples/zenoh_sub.cpp [hunk 24]
```diff
@@ -19,30 +21,41 @@
...
+}  // namespace
+
 auto main(int argc, const char* argv[]) -> int {
+  std::ignore = std::signal(SIGINT, [](int signal) {
+    if (signal == SIGINT) {
+      LOG(INFO) << "SIGINT received. Shutting down...";
+      keep_running.store(false);
```
Comment (filippobrizzi) [inline modules/examples/examples/zenoh_sub.cpp:32]:
```suggestion
      stop_flag.test_and_set();
      stop_flag.notify_all();
```

File: modules/examples/examples/zenoh_sub.cpp [hunk 25]
```diff
@@ -19,30 +21,41 @@
...
+          metadata.topic, metadata.sender_id, metadata.sequence_id, *pose);
     };
     auto subscriber = heph::ipc::subscribe<heph::ipc::zenoh::Subscriber, heph::examples::types::Pose>(
         session, std::move(topic_config), std::move(cb));
     (void)subscriber;
 
-    while (true) {
+    while (keep_running.load()) {
       std::this_thread::sleep_for(std::chrono::seconds{ 1 });
     }
```
Comment (filippobrizzi) [inline modules/examples/examples/zenoh_sub.cpp:60]:
```suggestion
    stop_flag.wait()
```

File: modules/ipc/apps/zenoh_string_service.cpp [hunk 26]
```diff
@@ -18,9 +18,11 @@ auto main(int argc, const char* argv[]) -> int {
 
     auto [config, topic_config] = parseArgs(args);
     auto session = heph::ipc::zenoh::createSession(std::move(config));
+    fmt::println("Simple service for std::string types for both Request and Reply. Don't use for services "
+                 "with different types.");
     fmt::println("Opening session: {}", heph::ipc::zenoh::toString(session->zenoh_session.info_zid()));
```
Comment (filippobrizzi) [inline modules/ipc/apps/zenoh_string_service.cpp:23]: Here we need `LOG(INFO)`

File: modules/ipc/apps/zenoh_string_service.cpp [hunk 27]
```diff
@@ -18,9 +18,11 @@ auto main(int argc, const char* argv[]) -> int {
 
     auto [config, topic_config] = parseArgs(args);
     auto session = heph::ipc::zenoh::createSession(std::move(config));
+    fmt::println("Simple service for std::string types for both Request and Reply. Don't use for services "
+                 "with different types.");
```
Comment (filippobrizzi) [inline modules/ipc/apps/zenoh_string_service.cpp:22]: This string can probably be the description of the program in line 14.

File: modules/ipc/include/hephaestus/ipc/zenoh/service.h [hunk 28]
```diff
@@ -4,30 +4,200 @@
...
+  if (const auto& error = std::get_if<zenohc::ErrorMessage>(&result)) {
+    LOG(ERROR) << fmt::format("Error on service reply for '{}': '{}'", topic_name, error->as_string_view());
+    return;
+  }
+
+  const auto& sample = std::get_if<zenohc::Sample>(&result);
+  DLOG(INFO) << fmt::format("Received answer of '{}'", sample->get_keyexpr().as_string_view());
+  if constexpr (std::is_same_v<ReplyT, std::string>) {
+    CHECK(sample->get_encoding() == Z_ENCODING_PREFIX_TEXT_PLAIN)
+        << "Encoding for std::string should be Z_ENCODING_PREFIX_TEXT_PLAIN";
```
Comment (filippobrizzi) [inline modules/ipc/include/hephaestus/ipc/zenoh/service.h:99]: should we throw an exception maybe? I feel we need to be consistent in error handling

File: modules/ipc/include/hephaestus/ipc/zenoh/service.h [hunk 29]
```diff
@@ -4,30 +4,200 @@
...
+}
+
+template <typename RequestT, typename ReplyT>
+auto callService(Session& session, const TopicConfig& topic_config, const RequestT& request,
+                 const std::optional<std::chrono::milliseconds>& timeout /*= std::nullopt*/)
+    -> std::vector<ServiceResponse<ReplyT>> {
+  internal::checkTemplatedTypes<RequestT, ReplyT>();
+
+  LOG(INFO) << fmt::format("Calling service on '{}'", topic_config.name);
+  std::mutex m;
```
Comment (filippobrizzi) [inline modules/ipc/include/hephaestus/ipc/zenoh/service.h:151]: nit.
```suggestion
  std::mutex mutex;
```

File: modules/ipc/include/hephaestus/ipc/zenoh/service.h [hunk 30]
```diff
@@ -4,30 +4,200 @@
...
+
+  const auto& sample = std::get_if<zenohc::Sample>(&result);
+  DLOG(INFO) << fmt::format("Received answer of '{}'", sample->get_keyexpr().as_string_view());
+  if constexpr (std::is_same_v<ReplyT, std::string>) {
+    CHECK(sample->get_encoding() == Z_ENCODING_PREFIX_TEXT_PLAIN)
+        << "Encoding for std::string should be Z_ENCODING_PREFIX_TEXT_PLAIN";
+    DLOG(INFO) << fmt::format("Payload is string: '{}'", sample->get_payload().as_string_view());
+    std::unique_lock<std::mutex> lock(m);
+    reply_messages.emplace_back(ServiceResponse<ReplyT>{
+        .topic = topic_name, .value = static_cast<std::string>(sample->get_payload().as_string_view()) });
```
Comment (filippobrizzi) [inline modules/ipc/include/hephaestus/ipc/zenoh/service.h:103]:
```suggestion
        .topic = sample.get_keyexpr().as_string_view(), .value = static_cast<std::string>(sample->get_payload().as_string_view()) });
```

File: modules/ipc/include/hephaestus/ipc/zenoh/service.h [hunk 31]
```diff
@@ -4,30 +4,200 @@
...
+  } else {
+    CHECK(sample->get_encoding() == Z_ENCODING_PREFIX_EMPTY)
+        << "Encoding for binary types should be Z_ENCODING_PREFIX_EMPTY";
+    auto payload = sample->get_payload();
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+    std::span<const std::byte> buffer(reinterpret_cast<const std::byte*>(payload.start), payload.len);
+    ReplyT reply_deserialized;
+    serdes::deserialize(buffer, reply_deserialized);
+    std::unique_lock<std::mutex> lock(m);
+    reply_messages.emplace_back(topic_name, std::move(reply_deserialized));
```
Comment (filippobrizzi) [inline modules/ipc/include/hephaestus/ipc/zenoh/service.h:113]:
```suggestion
    reply_messages.emplace_back(sample.get_keyexpr().as_string_view(), std::move(reply_deserialized));
```

File: modules/ipc/include/hephaestus/ipc/zenoh/service.h [hunk 32]
```diff
@@ -4,30 +4,200 @@
...
+    auto reply = this->callback_(internal::deserializeRequest<RequestT>(query));
+    zenohc::QueryReplyOptions options;
+    if constexpr (std::is_same_v<ReplyT, std::string>) {
+      options.set_encoding(zenohc::Encoding{ Z_ENCODING_PREFIX_TEXT_PLAIN });  // Update encoding.
+      query.reply(this->topic_config_.name, reply, options);
+    } else {
+      options.set_encoding(zenohc::Encoding{ Z_ENCODING_PREFIX_EMPTY });  // Update encoding.
+      auto buffer = serdes::serialize(reply);
+      DLOG(INFO) << "Replying with buffer of size: " << buffer.size();
+      DLOG(INFO) << fmt::format("Payload: {}", fmt::join(buffer, ","));
```
Comment (filippobrizzi) [inline modules/ipc/include/hephaestus/ipc/zenoh/service.h:135]:
```suggestion
      DLOG(INFO) << fmt::format("Reply: payload size: {}, content: {}", buffer.size(), fmt::join(buffer, ","));
```

File: modules/ipc/include/hephaestus/ipc/zenoh/service.h [hunk 33]
```diff
@@ -4,30 +4,200 @@
...
+  }
+
+  if constexpr (std::is_same_v<RequestT, std::string>) {
+    options.set_value(zenohc::Value(request, Z_ENCODING_PREFIX_TEXT_PLAIN));
+  } else {
+    auto buffer = serdes::serialize(request);
+    buffer = internal::addChangingBytes(buffer);
+
+    DLOG(INFO) << "Sending buffer of size: " << buffer.size();
+    DLOG(INFO) << fmt::format("Payload: {}", fmt::join(buffer, ","));
```
Comment (filippobrizzi) [inline modules/ipc/include/hephaestus/ipc/zenoh/service.h:176]:
```suggestion
    DLOG(INFO) << fmt::format("Request: payload size: {}, content: {}", buffer.size(), fmt::join(buffer, ","));
```

File: modules/ipc/include/hephaestus/ipc/zenoh/service.h [hunk 34]
```diff
@@ -4,30 +4,200 @@
...
+      return {};
+    }
+  } else {
+    done_signal.wait(lock);
+  }
+
+  if (!success) {
+    LOG(ERROR) << fmt::format("Error while calling service on '{}': '{}'", topic_config.name, error_code);
+    return {};
+  }
```
Comment (filippobrizzi) [inline modules/ipc/include/hephaestus/ipc/zenoh/service.h:198]: success returns immediately so this should be check right after the call with `throwExceptionIf`. 
Here we should check `error_code`

File: modules/ipc/include/hephaestus/ipc/zenoh/utils.h [hunk 35]
```diff
@@ -67,7 +67,7 @@ constexpr auto toMode(const zenohc::WhatAmI& me) -> Mode {
 
 inline auto toStringVector(const zenohc::StrArrayView& arr) -> std::vector<std::string> {
   const auto size = arr.get_len();
-  std::vector<std::string> vec;
+  std::vector<std::string> vec{};
```
Comment (filippobrizzi) [inline modules/ipc/include/hephaestus/ipc/zenoh/utils.h:70]: this is not needed, clang-tidy should actually give warning, no?

File: modules/ipc/src/topic_database.cpp [hunk 36]
```diff
@@ -40,10 +40,12 @@ auto ZenohTopicDatabase::getTypeInfo(const std::string& topic) -> const serdes::
...
   auto query_topic = getTypeInfoServiceTopic(topic);
 
-  const auto response = zenoh::query(session_->zenoh_session, query_topic, "");
+  static constexpr auto TIMEOUT = std::chrono::milliseconds{ 500 };
+  const auto response = zenoh::callService<std::string, std::string>(
+      *session_, TopicConfig{ .name = query_topic }, "", TIMEOUT);
   throwExceptionIf<heph::InvalidDataException>(
       response.size() != 1,
-      fmt::format("received {} responses for type from service {}", response.size(), query_topic));
+      fmt::format("received no or too many responses for type from service {}", query_topic));
```
Comment (filippobrizzi) [inline modules/ipc/src/topic_database.cpp:48]: why this change?

File: modules/examples/examples/zenoh_sub.cpp (hunk 23 above)
Comment (filippobrizzi) [reply modules/examples/examples/zenoh_sub.cpp:25]: have a look at `zenoh_topic_list.cpp`

File: modules/ipc/src/topic_database.cpp (hunk 36 above)
Comment (floriantschopp) [reply modules/ipc/src/topic_database.cpp:48]: was a remenance of some prev change, sorry

**Review by filippobrizzi — APPROVED** [review]
Fantastic!!
